docs/design/datacontracts/contract-descriptor.md

```diff
@@ -0,0 +1,151 @@
+and the [compatible contracts](./datacontracts_design.md#Compatible_Contract).
+
+The compatible contracts are stored in the toplevel key `"contracts"`.  The value will be a
+dictionary that contains each contract name as a key.  Each value is the version of the contract as
+a JSON integer constant.
+
+**Contract example**:
+
+``` jsonc
+{"Thread":1,"GCHandle":1,...}
+```
+
+**Complete in-memory data descriptor example**:
+
+``` jsonc
+{
+  "version": "0",
+  "baseline": "example-64",
+  "types":
+  {
+    "Thread": { "ThreadId": 32, "ThreadState": 0, "Next": 128 },
+    "ThreadStore": { "ThreadCount": 32, "ThreadList": 8 }
+  },
+  "globals":
+  {
+    "FEATURE_COMINTEROP": 0,
+    "s_pThreadStore": [ 0 ] // indirect from aux data offset 0
+  },
+  "contracts": {"Thread": 1,"GCHandle": 1, "ThreadStore": 1}
+}
+```
+
+## Contract symbol
+
+To aid in the discovery of the contract descriptor, the contract should be exported by the target
+process with the name `DotNetRuntimeContractDescriptor`.
+
+The meaning of the symbol differs on Windows and non-Windows platforms.
+
+### Windows
+
+Multiple DLLs loaded by a process may host a single .NET runtime.  Each DLL shall export the symbol
+`DotNetRuntimeContractDescriptor` pointing to a `struct DotNetRuntimeContractDescriptor`.  It is
+expected that `isList` will be 0.
+
+### Non-Windows
+
+In a process, each shared object containing a .NET runtime shall weakly-export the symbol
+`DotNetRuntimeContractDescriptor` (Using the C symbol conventions of the target platform.  That is,
+on platforms where such symbols typically have an `_` prepended, this symbol should be exported as
+`_DotNetRuntimeContractDescriptor`) with a null initial value.  As each .NET runtime in the process starts
+up, it shall atomically store a pointer to a `struct DotNetRuntimeContractDescriptorList` in
+`DotNetRuntimeContractDescriptor` where `next_runtime` points to the previous value of
+`DotNetRuntimeContractDescriptor` as by the C code below.
+
+The `runtime_name` is an arbitrary identifier to aid diagnostic tooling in identifying the current
+runtime.  (For example hosted runtimes may want to embed the name of the host; a desktop runtime may
+use just the runtime flavor and version)
+
+``` c
+typedef struct DotNetRuntimeContractDescriptorList* DescPtr;
+typedef _Atomic(DescPtr) AtomicDescPtr;
+
+// global weak symbol
+AtomicDescPtr  __attribute__((weak)) DotNetRuntimeContractDescriptor;
+
+static const struct DotNetRuntimeContractDescriptor g_private_descriptor = { ... }; // predefined descriptor for current runtime
+
+// to be called at startup
+void
+install_descriptor(const char *runtime_name)
+{
+    DescPtr descriptor = malloc(sizeof(struct DotNetRuntimeContractDescriptorList));
+    assert (descriptor != NULL);
+    descriptor->descriptor = g_private_descriptor; // copy the constant values
```

Comment:
Can this a be pointer instead of a copy?

Reply:
I may need to revisit this.  The issue is that we don't know the pointer size and endianness until we read off the magic and the flags.  

On the other hand that's somewhat untrue - since we're relying on the diagnostic tool's ability to find the addresses of symbols in the target process at some level the tool must already know what architecture it's looking at.

Reply:
Actually this whole list mechanism just doesn't work without knowing up front the target pointer size and endianness.

Is there really a scenario where the diagnostic tooling doesn't know the architecture of the target process?  Do minidumps have that info?

Reply:
SOS can deal with multiple runtimes in a process.  The `!runtimes` is the command to list and switch between them.  SOS/CLRMD (SOS uses CLRMD's runtime enumeration) currently does include any Native AOT modules. The internal DCR based SOS extensions does this Native AOT runtime enumeration. 

When there is enough cDAC support for Native AOT, we can change the SOS/CLRMD to include them.

Reply:
Removed the `DotNetRuntimeContractDescriptorList`.  clarified that for multi-runtime scenarios the diagnostic tooling should look in each loaded module for the contract descriptor